packages/ai/src/ai/modules/task/pipeline.py

```
@@ -0,0 +1,24 @@
+    Returns:
+        The source component ID, or None if no source component found.
+
+    Raises:
+        ValueError: If multiple source components are found.
+    """
+    source = None
+    for component in pipeline.get('components', []):
+        config = component.get('config', {})
+        if config.get('mode', '') == 'Source':
+            if source is not None:
+                raise ValueError('Pipeline has multiple source components, please specify one explicitly')
```

Ruff (0.15.7)

[warning] Avoid specifying long messages outside the exception class

(TRY003)